modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxManagerImpl.java

```diff
@@ -295,139 +319,252 @@ public void finishFull(HybridTimestampTracker timestampTracker, UUID txId, boole
-                                    updateTxMeta(txId, old -> {
-                                        if (isFinalState(old.txState())) {
-                                            finishingStateMeta.txFinishFuture().complete(old);
-
-                                            return old;
-                                        }
-
-                                        assert old instanceof TxStateMetaFinishing;
-
-                                        TxStateMeta finalTxStateMeta = coordinatorFinalTxStateMeta(verifiedCommit, commitTimestamp);
-
-                                        finishingStateMeta.txFinishFuture().complete(finalTxStateMeta);
-
-                                        return finalTxStateMeta;
-                                    });
-
-                                    if (verifiedCommit) {
-                                        observableTimestampTracker.update(commitTimestamp);
-                                    }
-                                });
-                    })
-                    .thenCompose(Function.identity())
-                    // verification future is added in order to share proper exception with the client
-                    .thenCompose(r -> verificationFuture);
-        };
-
-        AtomicReference<CompletableFuture<Void>> ref = new AtomicReference<>();
+        AtomicBoolean performingFinish = new AtomicBoolean();
         TxContext tuple = txCtxMap.compute(txId, (uuid, tuple0) -> {
             if (tuple0 == null) {
                 tuple0 = new TxContext(); // No writes enlisted.
             }
 
-            if (tuple0.finishFut == null) {
-                tuple0.finishFut = new CompletableFuture<>();
-                ref.set(tuple0.finishFut);
+            if (!tuple0.isTxFinishing()) {
+                tuple0.finishTx();
+
+                performingFinish.set(true);
             }
 
             return tuple0;
         });
 
-        if (ref.get() != null) { // This is a finishing thread.
-            if (!commit) {
-                clo.apply(null).handle((ignored, err) -> {
+        // This is a finishing thread.
+        if (performingFinish.get()) {
+            Function<Void, CompletableFuture<Void>> finishAction = ignored ->
+                    makeFinishRequest(
+                            observableTimestampTracker,
+                            commitPartition,
+                            commit,
+                            enlistedGroups,
+                            txId,
+                            finishingStateMeta.txFinishFuture()
+                    );
+
+            runFinish(commit, tuple, finishAction);
+        }
+
+        // The method `runFinish` has a side effect on `finishInProgressFuture` future, it kicks off another future that will complete it.
+        return tuple.finishInProgressFuture;
+    }
+
+    private static void runFinish(boolean commit, TxContext tuple, Function<Void, CompletableFuture<Void>> finishAction) {
```

Comment:
I like the idea of splitting the spaghetti code to the smaller methods, but their names should make sense in order to allow guessing which logic they contain. Now in this PR we have 5 "finish" methods:
- finish
- finishEmpty
- finishFull
- finishOnPrimary
- runFinish

and I am already lost...

Same with `makeFinishRequest` (for which making the request is not even its main job) and `makeDurableFinishRequest`.

I suggest to 
- remove `finishEmpty`, `runFinish` and `makeFinishRequest`;
- `finishOnPrimary` should be renamed to `makeFinishRequest`;
- move the logic related to computing of `performingFinish` to a method inside of TxContext;
- rename `makeDurableFinishRequest` to `durableFinish`

WDYT?